mlir/lib/Transforms/lab4/nogin_denis/FusedMultAddPass.cpp

```diff
@@ -0,0 +1,64 @@
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+
+using namespace mlir;
+
+namespace {
+class FusedMultAddPass
+    : public PassWrapper<FusedMultAddPass, OperationPass<ModuleOp>> {
+public:
+  StringRef getArgument() const final { return "fused-mult-add"; }
+  StringRef getDescription() const final {
+    return "Fuses multiply and add operations into a single fma operation";
+  }
+
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+
+    module.walk([&](Operation *op) {
+      if (auto addOp = dyn_cast<LLVM::FAddOp>(op)) {
+        Value addLHS = addOp.getOperand(0);
+        Value addRHS = addOp.getOperand(1);
+
+        auto tryFuse = [&](Value mulOperand, Value otherOperand, Value &fma) {
+          if (auto mulOp = mulOperand.getDefiningOp<LLVM::FMulOp>()) {
+            OpBuilder builder(addOp);
+            fma =
+                builder.create<LLVM::FMAOp>(addOp.getLoc(), mulOp.getOperand(0),
+                                            mulOp.getOperand(1), otherOperand);
+            addOp.replaceAllUsesWith(fma);
+            return true;
+          }
+          return false;
+        };
+
+        Value fma;
+        if (tryFuse(addLHS, addRHS, fma) || tryFuse(addRHS, addLHS, fma)) {
```

Comment:
I'm curious, why do we pass this "fma" value by reference and then set it to created value and then don't use that value?

Reply:
Yes, you're right. When I wrote this, I thought it would be necessary in the future. I'll fix it now.